src/jsonrpc.c

```diff
@@ -0,0 +1,493 @@
+	return (status);
+}
+
+static int
+close_response(yajl_gen genctx)
+{
+	int	status;
+
+	if (YAJL_GEN_KO(status = yajl_gen_map_close(genctx)))
+		goto failed;
+	status = yajl_gen_map_close(genctx);
+failed:
+	return (status);
+}
+
+static int
+write_log(struct jsonrpc_request *req)
+{
+	bool	wrote_smth = false;
+	int	status = 0;
+	
+	for (struct jsonrpc_log *log = req->log.next; log != &req->log;
+		log = log->next) {
+
+		if (((1 << log->lvl) & req->log_levels) == 0)
+			continue;
+
+		if (!wrote_smth) {
+			if (YAJL_GEN_KO(status = YAJL_GEN_CONST_STRING(req->gen,
+			    "data")))
+				goto failed;
+			if (YAJL_GEN_KO(status = yajl_gen_array_open(req->gen)))
+				goto failed;
+			yajl_gen_config(req->gen, yajl_gen_validate_utf8, 1);
+			wrote_smth = true;
+		}
+		
+		if (YAJL_GEN_KO(status = yajl_gen_array_open(req->gen)))
+			goto failed;
+		if (YAJL_GEN_KO(status = yajl_gen_integer(req->gen, log->lvl)))
+			goto failed;
+		if (YAJL_GEN_KO(status = yajl_gen_string(req->gen,
+		    (unsigned char *)log->msg, strlen(log->msg))))
+			goto failed;
+		if (YAJL_GEN_KO(status = yajl_gen_array_close(req->gen)))
+			goto failed;
+	}
+
+	if (wrote_smth) {
+		yajl_gen_config(req->gen, yajl_gen_validate_utf8, 0);
+		status = yajl_gen_array_close(req->gen);
+	}
+failed:
+	return (status);
+}
+
+static int
+write_error(struct jsonrpc_request *req, int code, const char *message)
+{
+	int	status;
+
+	yajl_gen_config(req->gen, yajl_gen_validate_utf8, 0);
+
+	if (YAJL_GEN_KO(status = open_response(req->gen, req->id)))
+		goto failed;
+	if (YAJL_GEN_KO(status = YAJL_GEN_CONST_STRING(req->gen, "error")))
+		goto failed;
+	if (YAJL_GEN_KO(status = yajl_gen_map_open(req->gen)))
+		goto failed;
+	if (YAJL_GEN_KO(status = YAJL_GEN_CONST_STRING(req->gen, "code")))
+		goto failed;
+	if (YAJL_GEN_KO(status = yajl_gen_integer(req->gen, code)))
+		goto failed;
+	if (YAJL_GEN_KO(status = YAJL_GEN_CONST_STRING(req->gen, "message")))
+		goto failed;
+
+	yajl_gen_config(req->gen, yajl_gen_validate_utf8, 1);
+
+	if (YAJL_GEN_KO(status = yajl_gen_string(req->gen,
+			(const unsigned char *)message, strlen(message))))
+		goto failed;
+
+	yajl_gen_config(req->gen, yajl_gen_validate_utf8, 0);
+
+	if (YAJL_GEN_KO(status = write_log(req)))
+		goto failed;
+
+	status = close_response(req->gen);
+failed:
+	return (status);
+}
+
+static const char *
+known_msg(int code)
+{
+	switch (code) {
+	case JSONRPC_PARSE_ERROR:
+		return (JSONRPC_PARSE_ERROR_MSG);
+	case JSONRPC_INVALID_REQUEST:
+		return (JSONRPC_INVALID_REQUEST_MSG);
+	case JSONRPC_METHOD_NOT_FOUND:
+		return (JSONRPC_METHOD_NOT_FOUND_MSG);
+	case JSONRPC_INVALID_PARAMS:
+		return (JSONRPC_INVALID_PARAMS_MSG);
+	case JSONRPC_INTERNAL_ERROR:
+		return (JSONRPC_INTERNAL_ERROR_MSG);
+	case JSONRPC_SERVER_ERROR:
+		return (JSONRPC_SERVER_ERROR_MSG);
+	case JSONRPC_LIMIT_REACHED:
+		return (JSONRPC_LIMIT_REACHED_MSG);
+	default:
+		return (NULL);
+	}
+}
+
+int
+jsonrpc_error(struct jsonrpc_request *req, int code, const char *msg)
```

Comment:
The initialization as well is invisible to the application. Only the allocation / deallocation. In standard HTTP request / response cycle, after having responded to a request, whether with a result or an error, there's no point anymore to output something else. Nothing more should be appended really.

I'll anyway write a version with destruction up to the caller. I think it will take some more days and change bits here and there.